v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+/*DAD29F36-0A48-4DEF-9D50-8EF9036B92B4*/
+
+/* Codes_SRS_LOG_SINK_ETW_01_002: [ log_sink_etw_log shall maintain the state of whether TraceLoggingRegister was called in a variable accessed via InterlockedXXX APIs, which shall have 3 possible values: NOT_REGISTERED (1), REGISTERING (2), REGISTERED(3). ]*/
+#define PROVIDER_STATE_VALUES \
+    PROVIDER_STATE_NOT_REGISTERED, \
+    PROVIDER_STATE_REGISTERING, \
+    PROVIDER_STATE_REGISTERED \
+
+MU_DEFINE_ENUM(PROVIDER_STATE, PROVIDER_STATE_VALUES)
+
+static volatile LONG etw_provider_state = PROVIDER_STATE_NOT_REGISTERED;
+
+static void log_sink_etw_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line, const char* message_format, ...);
+
+static void internal_log_sink_etw_lazy_register_provider(void)
+{
+    LONG state;
+
+    /* Codes_SRS_LOG_SINK_ETW_01_003: [ log_sink_etw_log shall perform the below actions until the provider is registered or an error is encountered: ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_004: [ If the state is NOT_REGISTERED (1): ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_005: [ log_sink_etw_log shall switch the state to REGISTERING (2). ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_009: [ Checking and changing the variable that maintains whether TraceLoggingRegister was called shall be done using InterlockedCompareExchange and InterlockedExchange. ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_011: [ If the state is REGISTERED (3), log_sink_etw_log shall proceed to log the ETW event. ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_087: [ If the state is REGISTERING (2) log_sink_etw_log shall wait until the state is not REGISTERING (2). ]*/
+    while ((state = InterlockedCompareExchange(&etw_provider_state, PROVIDER_STATE_REGISTERING, PROVIDER_STATE_NOT_REGISTERED)) != PROVIDER_STATE_REGISTERED)
+    {
+        if (state == PROVIDER_STATE_NOT_REGISTERED)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_006: [ log_sink_etw_log shall register the ETW TraceLogging provider by calling TraceLoggingRegister (TraceLoggingRegister_EventRegister_EventSetInformation). ]*/
+            TLG_STATUS register_result = TraceLoggingRegister(g_my_component_provider);
+            if (SUCCEEDED(register_result))
+            {
+                /* Codes_SRS_LOG_SINK_ETW_01_007: [ log_sink_etw_log shall switch the state to REGISTERED. ]*/
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_REGISTERED);
+                /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw_log shall printf the fact that the provider was registered and from which executable (as obtained by calling _pgmptr). ]*/
+
+                char* executable_full_path_name;
+                if (_get_pgmptr(&executable_full_path_name) != 0)
+                {
+                    /* Codes_SRS_LOG_SINK_ETW_01_083: [ If _get_pgmptr fails, the executable shall be printed as UNKNOWN. ]*/
+                    executable_full_path_name = "UNKNOWN";
```

Comment:
>executable_full_path_name = "UNKNOWN"; [](http://example.com/codeflow?start=20&length=38)

How did this compile?

Reply:
Compiles just fine actually.
And it is valid.

Example 8 on page 144 (section 6.7.9 Initialization) in C11 shows exactly this case.

However, it would be UB if this would get modified. So in order to avoid some confusion I introduced a const char*